ipld/src/selector/mod.rs

```diff
@@ -236,9 +235,126 @@ impl Selector {
+                    Ipld::List(l) => {
+                        let i = p.to_index()?;
+                        // Check to make sure index is within list bounds
+                        if i < start || i >= end || i >= l.len() {
+                            None
+                        } else {
+                            // Path segment is within the selector range
+                            Some(*next)
+                        }
+                    }
+                    _ => None,
+                }
+            }
+            ExploreRecursive {
+                current,
+                sequence,
+                limit,
+                stop_at,
+            } => {
+                let next = match current {
+                    Some(s) => s.explore(ipld, p)?,
+                    None => sequence.clone().explore(ipld, p)?,
+                };
+
+                if !has_recursive_edge(&next) {
+                    return Some(ExploreRecursive {
+                        sequence,
+                        current: Some(next.into()),
+                        limit,
+                        stop_at,
+                    });
+                }
+                match limit {
+                    RecursionLimit::Depth(depth) => {
+                        if depth < 2 {
+                            // Replaces recursive edge with None on last iteration
+                            // TODO revisit, shouldn't need to replace, would be better to just
+                            // return none when edge is hit on final depth
+                            return replace_recursive_edge(next, None).map(|s| *s);
+                        }
+                        Some(ExploreRecursive {
+                            current: replace_recursive_edge(next, Some(sequence.clone())),
+                            sequence,
+                            limit: RecursionLimit::Depth(depth - 1),
+                            stop_at,
+                        })
+                    }
+                    RecursionLimit::None => Some(ExploreRecursive {
+                        current: replace_recursive_edge(next, Some(sequence.clone())),
+                        sequence,
+                        limit,
+                        stop_at,
+                    }),
+                }
+            }
+            ExploreUnion(selectors) => {
+                let mut replace_selectors = Vec::with_capacity(selectors.len());
+                for s in selectors {
+                    if let Some(new) = s.explore(ipld, p) {
+                        // Push all explorable selectors in union
+                        replace_selectors.push(new)
+                    }
+                }
+                if replace_selectors.is_empty() {
+                    return None;
+                }
+                if replace_selectors.len() == 1 {
+                    return Some(replace_selectors.pop().unwrap());
+                }
+                Some(ExploreUnion(replace_selectors))
```

Comment:
You have this twice, you could clean this up with an associated function that does it for you:
```rust
fn from_vec(mut vec: Vec<Self>) -> Option<Self> {
    match vec.len() {
        0 | 1 => vec.pop(),
        _ => Some(ExploreUnion(vec)),
    }
}
```

Reply:
that's a good idea, I will modify because that will panic on 0 and I think it might make sense to have a `From<Vec<Selector>>` impl instead (I'll think about it a bit more before doing)